### PR TITLE
[PyTorch] Port ExecuTorch bfdot improvement back to ATen BlasKernel

### PR DESCRIPTION
**Pull request by swolchok**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #136331
* #136445

ExecuTorch's fork of BlasKernel.cpp grew bfdot support, complete with demonstration that it helps. Port it back to PyTorch. Supersedes https://github.com/pytorch/pytorch/pull/127488 . Includes https://github.com/pytorch/executorch/pull/5444 .

Differential Revision: [D63045939](https://our.internmc.facebook.com/intern/diff/D63045939/)

### CODE REVIEW

**Comment by swolchok** [comment]
Note that this code is gated off under C10_MOBILE, but we have ExecuTorch for that anyway, right?

File: aten/src/ATen/native/BlasKernel.cpp [hunk 1]
```diff
@@ -510,12 +591,31 @@ float dot_with_fp32_arith(const T* vec1, const T* vec2, int64_t len) {
   return reducedSum;
 }
 
+TARGET_ARM_BF16_ATTRIBUTE double
```
Comment (malfet) [inline aten/src/ATen/native/BlasKernel.cpp:594]: Why double here?
```suggestion
TARGET_ARM_BF16_ATTRIBUTE float
```

File: aten/src/ATen/native/BlasKernel.cpp [hunk 2]
```diff
@@ -510,12 +591,31 @@ float dot_with_fp32_arith(const T* vec1, const T* vec2, int64_t len) {
...
+dot_with_fp32_arith_bfdot(const BFloat16* vec1, const BFloat16* vec2, int64_t len) {
+  auto reducedSum = dot_with_fp32_arith_main_loop_bfdot(vec1, vec2, len);
+  return dot_with_fp32_arith_tail_after_main_loop(vec1, vec2, len, reducedSum);
+}
+
+template <typename T>
+C10_ALWAYS_INLINE double
```
Comment (malfet) [inline aten/src/ATen/native/BlasKernel.cpp:601]:
```suggestion
C10_ALWAYS_INLINE float
```

File: c10/macros/Macros.h [hunk 3]
```diff
@@ -244,6 +244,14 @@ using namespace c10::xpu;
 #define C10_ALWAYS_INLINE inline
 #endif
 
+#if defined(_MSC_VER)
+#define C10_ALWAYS_INLINE_ATTRIBUTE
+#elif __has_attribute(always_inline) || defined(__GNUC__)
+#define C10_ALWAYS_INLINE_ATTRIBUTE __attribute__((__always_inline__))
+#else
+#define C10_ALWAYS_INLINE_ATTRIBUTE
+#endif
```
Comment (malfet) [inline c10/macros/Macros.h:253]: Can you please move it to a separate PR? Also, why is this needed?(any perf numbers) And why exclude clang?

Comment (swolchok) [reply c10/macros/Macros.h:253]: > why is this needed?(any perf numbers)

not having it prevents ForcedUnroll from working properly under -Oz. See https://github.com/pytorch/executorch/pull/5247 .

> why exclude clang

clang defines __GNUC__.

Comment (swolchok) [reply c10/macros/Macros.h]: > separate PR

done. https://github.com/pytorch/pytorch/pull/136445

**Review by malfet — APPROVED** [review]

File: aten/src/ATen/native/BlasKernel.cpp [hunk 4]
```diff
@@ -455,18 +456,61 @@ static C10_ALWAYS_INLINE float32x4_t f32_fma_bf16(float32x4_t a, uint16x4_t b, u
...
-  const uint16x8_t temp_vec2 = vld1q_u16(reinterpret_cast<const uint16_t*>(&vec2[registerPairIndex * 2 * kF32ElementsPerRegister]));
+#if defined(__clang__) && __clang_major__ > 15
+// https://godbolt.org/z/z8P4Yncra
+#define COMPILER_SUPPORTS_BF16_TARGET 1
+#elif !defined(__clang__) && defined(__GNUC__) && __GNUC__ >= 10
+// https://gcc.gnu.org/gcc-10/changes.html
+// https://godbolt.org/z/cdGG7vn8o
+#define COMPILER_SUPPORTS_BF16_TARGET 1
+#else
+#define COMPILER_SUPPORTS_BF16_TARGET 0
```
Comment (albanD) [inline aten/src/ATen/native/BlasKernel.cpp:467]: Doesn't this override the definitions above?

File: aten/src/ATen/native/BlasKernel.cpp [hunk 5]
```diff
@@ -301,7 +302,7 @@ static constexpr auto kF16RegistersPerIterationShift = kF16ElementsPerIterationS
 static constexpr auto kF16RegistersPerIteration = 1 << kF16RegistersPerIterationShift;
 static_assert(kF16RegistersPerIteration == kF16ElementsPerIteration / kF16ElementsPerRegister);
 
-static inline double reduce(float16x8_t x[kF16RegistersPerIteration]) {
+static inline float reduce(float16x8_t x[kF16RegistersPerIteration]) {
```
Comment (albanD) [inline aten/src/ATen/native/BlasKernel.cpp:305]: Wasn't this done on purpose to preserve accuracy?

**Review by albanD — CHANGES_REQUESTED** [review]
CI failure look very much relevant!

File: aten/src/ATen/native/BlasKernel.cpp (hunk 5 above)
Comment (swolchok) [reply aten/src/ATen/native/BlasKernel.cpp:305]: I believe @malfet asked for it. perhaps I got confused though.

File: aten/src/ATen/native/BlasKernel.cpp (hunk 4 above)
Comment (swolchok) [reply aten/src/ATen/native/BlasKernel.cpp:467]: it's in an #if/#elif/#else, so no.

Comment (albanD) [reply aten/src/ATen/native/BlasKernel.cpp:467]: Ho sorry, my bad

File: aten/src/ATen/native/BlasKernel.cpp (hunk 5 above)
Comment (albanD) [reply aten/src/ATen/native/BlasKernel.cpp:305]: Ok, in any case, the numerical error on macos arm seem relevant and needs investigation right?

Comment (swolchok) [reply aten/src/ATen/native/BlasKernel.cpp:305]: > Ok, in any case, the numerical error on macos arm seem relevant and needs investigation right?

yes, it's fixed. I accidentally replaced the float16 kernel with the bfloat16 kernel due to name changes

**Review by albanD — APPROVED** [review]
If ci says numerical match, that sounds ok to me.

**Comment by abhishek-iitmadras** [comment]
Hi @malfet , @albanD , @swolchok 

I am getting the following error while building pytorch on aarch64 :
```
[4792/6243] Building CXX object caffe2/CMakeFiles/torch_cpu.dir/__/aten/src/ATen/native/BlasKernel.cpp.o
FAILED: caffe2/CMakeFiles/torch_cpu.dir/__/aten/src/ATen/native/BlasKernel.cpp.o 
/usr/bin/ccache /usr/bin/g++-11 -DAT_BUILD_ARM_VEC256_WITH_SLEEF -DAT_PER_OPERATOR_HEADERS -DCAFFE2_BUILD_MAIN_LIB -DCPUINFO_SUPPORTED_PLATFORM=1 -DFLASHATTENTION_DISABLE_ALIBI -DFMT_HEADER_ONLY=1 -DFXDIV_USE_INLINE_ASSEMBLY=0 -DHAVE_MALLOC_USABLE_SIZE=1 -DHAVE_MMAP=1 -DHAVE_SHM_OPEN=1 -DHAVE_SHM_UNLINK=1 -DMINIZ_DISABLE_ZIP_READER_CRC32_CHECKS -DNNP_CONVOLUTION_ONLY=0 -DNNP_INFERENCE_ONLY=0 -DONNXIFI_ENABLE_EXT=1 -DONNX_ML=1 -DONNX_NAMESPACE=onnx_torch -DUSE_C10D_GLOO -DUSE_C10D_MPI -DUSE_DISTRIBUTED -DUSE_EXTERNAL_MZCRC -DUSE_RPC -DUSE_TENSORPIPE -D_FILE_OFFSET_BITS=64 -Dtorch_cpu_EXPORTS -I/home/abhishekk/org11/venv/pytorch/build/aten/src -I/home/abhishekk/org11/venv/pytorch/aten/src -I/home/abhishekk/org11/venv/pytorch/build -I/home/abhishekk/org11/venv/pytorch -I/home/abhishekk/org11/venv/pytorch/cmake/../third_party/benchmark/include -I/home/abhishekk/org11/venv/pytorch/third_party/onnx -I/home/abhishekk/org11/venv/pytorch/build/third_party/onnx -I/home/abhishekk/org11/venv/pytorch/nlohmann -I/home/abhishekk/org11/venv/pytorch/torch/csrc/api -I/home/abhishekk/org11/venv/pytorch/torch/csrc/api/include -I/home/abhishekk/org11/venv/pytorch/caffe2/aten/src/TH -I/home/abhishekk/org11/venv/pytorch/build/caffe2/aten/src/TH -I/home/abhishekk/org11/venv/pytorch/build/caffe2/aten/src -I/home/abhishekk/org11/venv/pytorch/build/caffe2/../aten/src -I/home/abhishekk/org11/venv/pytorch/torch/csrc -I/home/abhishekk/org11/venv/pytorch/third_party/miniz-2.1.0 -I/home/abhishekk/org11/venv/pytorch/third_party/kineto/libkineto/include -I/home/abhishekk/org11/venv/pytorch/third_party/kineto/libkineto/src -I/home/abhishekk/org11/venv/pytorch/third_party/cpp-httplib -I/home/abhishekk/org11/venv/pytorch/aten/src/ATen/.. -I/home/abhishekk/org11/venv/pytorch/third_party/FXdiv/include -I/home/abhishekk/org11/venv/pytorch/c10/.. -I/home/abhishekk/org11/venv/pytorch/third_party/pthreadpool/include -I/home/abhishekk/org11/venv/pytorch/third_party/cpuinfo/include -I/home/abhishekk/org11/venv/pytorch/aten/src/ATen/native/quantized/cpu/qnnpack/include -I/home/abhishekk/org11/venv/pytorch/aten/src/ATen/native/quantized/cpu/qnnpack/src -I/home/abhishekk/org11/venv/pytorch/aten/src/ATen/native/quantized/cpu/qnnpack/deps/clog/include -I/home/abhishekk/org11/venv/pytorch/third_party/NNPACK/include -I/home/abhishekk/org11/venv/pytorch/third_party/FP16/include -I/home/abhishekk/org11/venv/pytorch/third_party/tensorpipe -I/home/abhishekk/org11/venv/pytorch/build/third_party/tensorpipe -I/home/abhishekk/org11/venv/pytorch/third_party/tensorpipe/third_party/libnop/include -I/home/abhishekk/org11/venv/pytorch/third_party/fmt/include -I/home/abhishekk/org11/venv/pytorch/third_party/flatbuffers/include -isystem /home/abhishekk/org11/venv/pytorch/build/third_party/gloo -isystem /home/abhishekk/org11/venv/pytorch/cmake/../third_party/gloo -isystem /home/abhishekk/org11/venv/pytorch/cmake/../third_party/tensorpipe/third_party/libuv/include -isystem /home/abhishekk/org11/venv/pytorch/cmake/../third_party/googletest/googlemock/include -isystem /home/abhishekk/org11/venv/pytorch/cmake/../third_party/googletest/googletest/include -isystem /home/abhishekk/org11/venv/pytorch/third_party/protobuf/src -isystem /home/abhishekk/org11/venv/pytorch/third_party/XNNPACK/include -isystem /home/abhishekk/org11/venv/pytorch/cmake/../third_party/eigen -isystem /home/abhishekk/org11/venv/pytorch/INTERFACE -isystem /home/abhishekk/org11/venv/pytorch/third_party/nlohmann/include -isystem /home/abhishekk/org11/venv/pytorch/build/include -isystem /usr/lib/aarch64-linux-gnu/openmpi/include -isystem /usr/lib/aarch64-linux-gnu/openmpi/include/openmpi -D_GLIBCXX_USE_CXX11_ABI=1 -fvisibility-inlines-hidden -DUSE_PTHREADPOOL -DNDEBUG -DUSE_KINETO -DLIBKINETO_NOCUPTI -DLIBKINETO_NOROCTRACER -DLIBKINETO_NOXPUPTI=ON -DUSE_PYTORCH_QNNPACK -DAT_BUILD_ARM_VEC256_WITH_SLEEF -DUSE_XNNPACK -DSYMBOLICATE_MOBILE_DEBUG_HANDLE -O2 -fPIC -Wall -Wextra -Werror=return-type -Werror=non-virtual-dtor -Werror=range-loop-construct -Werror=bool-operation -Wnarrowing -Wno-missing-field-initializers -Wno-type-limits -Wno-array-bounds -Wno-unknown-pragmas -Wno-unused-parameter -Wno-strict-overflow -Wno-strict-aliasing -Wno-stringop-overflow -Wsuggest-override -Wno-psabi -Wno-error=old-style-cast -Wno-missing-braces -fdiagnostics-color=always -faligned-new -Wno-unused-but-set-variable -Wno-maybe-uninitialized -fno-math-errno -fno-trapping-math -Werror=format -Wno-stringop-overflow -DHAVE_SVE_CPU_DEFINITION -DHAVE_SVE256_CPU_DEFINITION -O3 -DNDEBUG -DNDEBUG -std=gnu++17 -fPIC -DTORCH_USE_LIBUV -DCAFFE2_USE_GLOO -D__NEON__ -Wall -Wextra -Wdeprecated -Wno-unused-parameter -Wno-missing-field-initializers -Wno-type-limits -Wno-array-bounds -Wno-unknown-pragmas -Wno-strict-overflow -Wno-strict-aliasing -Wunused-function -Wunused-variable -Wunused-but-set-variable -Wno-maybe-uninitialized -fvisibility=hidden -O2 -pthread -fopenmp -march=armv8.2-a+fp16 -MD -MT caffe2/CMakeFiles/torch_cpu.dir/__/aten/src/ATen/native/BlasKernel.cpp.o -MF caffe2/CMakeFiles/torch_cpu.dir/__/aten/src/ATen/native/BlasKernel.cpp.o.d -o caffe2/CMakeFiles/torch_cpu.dir/__/aten/src/ATen/native/BlasKernel.cpp.o -c /home/abhishekk/org11/venv/pytorch/aten/src/ATen/native/BlasKernel.cpp
/home/abhishekk/org11/venv/pytorch/aten/src/ATen/native/BlasKernel.cpp: In function ‘float at::native::blas_impl::{anonymous}::dot_with_fp32_arith_bfdot(const c10::BFloat16*, const c10::BFloat16*, int64_t)’:
/home/abhishekk/org11/venv/pytorch/aten/src/ATen/native/BlasKernel.cpp:582:25: error: inlining failed in call to ‘always_inline’ ‘float at::native::blas_impl::{anonymous}::dot_with_fp32_arith_tail_after_main_loop(const T*, const T*, int64_t, float) [with T = c10::BFloat16]’: target specific option mismatch
  582 | C10_ALWAYS_INLINE float dot_with_fp32_arith_tail_after_main_loop(
      |                         ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/home/abhishekk/org11/venv/pytorch/aten/src/ATen/native/BlasKernel.cpp:610:50: note: called from here
  610 |   return dot_with_fp32_arith_tail_after_main_loop(vec1, vec2, len, reducedSum);
      |          ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/home/abhishekk/org11/venv/pytorch/aten/src/ATen/native/BlasKernel.cpp:582:25: error: inlining failed in call to ‘always_inline’ ‘float at::native::blas_impl::{anonymous}::dot_with_fp32_arith_tail_after_main_loop(const T*, const T*, int64_t, float) [with T = c10::BFloat16]’: target specific option mismatch
  582 | C10_ALWAYS_INLINE float dot_with_fp32_arith_tail_after_main_loop(
      |                         ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/home/abhishekk/org11/venv/pytorch/aten/src/ATen/native/BlasKernel.cpp:610:50: note: called from here
  610 |   return dot_with_fp32_arith_tail_after_main_loop(vec1, vec2, len, reducedSum);
      |          ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~
```

**Comment by albanD** [comment]
@pytorchbot revert -m "Breaks aarch64 builds, see link below" -c nosignal

This also broke the inductor aarch64 builds: https://github.com/pytorch/pytorch/actions/runs/11175546054/job/31067129185

**Comment by swolchok** [comment]
> Hi @malfet , @albanD , @swolchok
> 
> I am getting the following error while building pytorch on aarch64 :
> 
> ```
> [4792/6243] Building CXX object caffe2/CMakeFiles/torch_cpu.dir/__/aten/src/ATen/native/BlasKernel.cpp.o
> FAILED: caffe2/CMakeFiles/torch_cpu.dir/__/aten/src/ATen/native/BlasKernel.cpp.o 
> /usr/bin/ccache /usr/bin/g++-11 -DAT_BUILD_ARM_VEC256_WITH_SLEEF -DAT_PER_OPERATOR_HEADERS -DCAFFE2_BUILD_MAIN_LIB -DCPUINFO_SUPPORTED_PLATFORM=1 -DFLASHATTENTION_DISABLE_ALIBI -DFMT_HEADER_ONLY=1 -DFXDIV_USE_INLINE_ASSEMBLY=0 -DHAVE_MALLOC_USABLE_SIZE=1 -DHAVE_MMAP=1 -DHAVE_SHM_OPEN=1 -DHAVE_SHM_UNLINK=1 -DMINIZ_DISABLE_ZIP_READER_CRC32_CHECKS -DNNP_CONVOLUTION_ONLY=0 -DNNP_INFERENCE_ONLY=0 -DONNXIFI_ENABLE_EXT=1 -DONNX_ML=1 -DONNX_NAMESPACE=onnx_torch -DUSE_C10D_GLOO -DUSE_C10D_MPI -DUSE_DISTRIBUTED -DUSE_EXTERNAL_MZCRC -DUSE_RPC -DUSE_TENSORPIPE -D_FILE_OFFSET_BITS=64 -Dtorch_cpu_EXPORTS -I/home/abhishekk/org11/venv/pytorch/build/aten/src -I/home/abhishekk/org11/venv/pytorch/aten/src -I/home/abhishekk/org11/venv/pytorch/build -I/home/abhishekk/org11/venv/pytorch -I/home/abhishekk/org11/venv/pytorch/cmake/../third_party/benchmark/include -I/home/abhishekk/org11/venv/pytorch/third_party/onnx -I/home/abhishekk/org11/venv/pytorch/build/third_party/onnx -I/home/abhishekk/org11/venv/pytorch/nlohmann -I/home/abhishekk/org11/venv/pytorch/torch/csrc/api -I/home/abhishekk/org11/venv/pytorch/torch/csrc/api/include -I/home/abhishekk/org11/venv/pytorch/caffe2/aten/src/TH -I/home/abhishekk/org11/venv/pytorch/build/caffe2/aten/src/TH -I/home/abhishekk/org11/venv/pytorch/build/caffe2/aten/src -I/home/abhishekk/org11/venv/pytorch/build/caffe2/../aten/src -I/home/abhishekk/org11/venv/pytorch/torch/csrc -I/home/abhishekk/org11/venv/pytorch/third_party/miniz-2.1.0 -I/home/abhishekk/org11/venv/pytorch/third_party/kineto/libkineto/include -I/home/abhishekk/org11/venv/pytorch/third_party/kineto/libkineto/src -I/home/abhishekk/org11/venv/pytorch/third_party/cpp-httplib -I/home/abhishekk/org11/venv/pytorch/aten/src/ATen/.. -I/home/abhishekk/org11/venv/pytorch/third_party/FXdiv/include -I/home/abhishekk/org11/venv/pytorch/c10/.. -I/home/abhishekk/org11/venv/pytorch/third_party/pthreadpool/include -I/home/abhishekk/org11/venv/pytorch/third_party/cpuinfo/include -I/home/abhishekk/org11/venv/pytorch/aten/src/ATen/native/quantized/cpu/qnnpack/include -I/home/abhishekk/org11/venv/pytorch/aten/src/ATen/native/quantized/cpu/qnnpack/src -I/home/abhishekk/org11/venv/pytorch/aten/src/ATen/native/quantized/cpu/qnnpack/deps/clog/include -I/home/abhishekk/org11/venv/pytorch/third_party/NNPACK/include -I/home/abhishekk/org11/venv/pytorch/third_party/FP16/include -I/home/abhishekk/org11/venv/pytorch/third_party/tensorpipe -I/home/abhishekk/org11/venv/pytorch/build/third_party/tensorpipe -I/home/abhishekk/org11/venv/pytorch/third_party/tensorpipe/third_party/libnop/include -I/home/abhishekk/org11/venv/pytorch/third_party/fmt/include -I/home/abhishekk/org11/venv/pytorch/third_party/flatbuffers/include -isystem /home/abhishekk/org11/venv/pytorch/build/third_party/gloo -isystem /home/abhishekk/org11/venv/pytorch/cmake/../third_party/gloo -isystem /home/abhishekk/org11/venv/pytorch/cmake/../third_party/tensorpipe/third_party/libuv/include -isystem /home/abhishekk/org11/venv/pytorch/cmake/../third_party/googletest/googlemock/include -isystem /home/abhishekk/org11/venv/pytorch/cmake/../third_party/googletest/googletest/include -isystem /home/abhishekk/org11/venv/pytorch/third_party/protobuf/src -isystem /home/abhishekk/org11/venv/pytorch/third_party/XNNPACK/include -isystem /home/abhishekk/org11/venv/pytorch/cmake/../third_party/eigen -isystem /home/abhishekk/org11/venv/pytorch/INTERFACE -isystem /home/abhishekk/org11/venv/pytorch/third_party/nlohmann/include -isystem /home/abhishekk/org11/venv/pytorch/build/include -isystem /usr/lib/aarch64-linux-gnu/openmpi/include -isystem /usr/lib/aarch64-linux-gnu/openmpi/include/openmpi -D_GLIBCXX_USE_CXX11_ABI=1 -fvisibility-inlines-hidden -DUSE_PTHREADPOOL -DNDEBUG -DUSE_KINETO -DLIBKINETO_NOCUPTI -DLIBKINETO_NOROCTRACER -DLIBKINETO_NOXPUPTI=ON -DUSE_PYTORCH_QNNPACK -DAT_BUILD_ARM_VEC256_WITH_SLEEF -DUSE_XNNPACK -DSYMBOLICATE_MOBILE_DEBUG_HANDLE -O2 -fPIC -Wall -Wextra -Werror=return-type -Werror=non-virtual-dtor -Werror=range-loop-construct -Werror=bool-operation -Wnarrowing -Wno-missing-field-initializers -Wno-type-limits -Wno-array-bounds -Wno-unknown-pragmas -Wno-unused-parameter -Wno-strict-overflow -Wno-strict-aliasing -Wno-stringop-overflow -Wsuggest-override -Wno-psabi -Wno-error=old-style-cast -Wno-missing-braces -fdiagnostics-color=always -faligned-new -Wno-unused-but-set-variable -Wno-maybe-uninitialized -fno-math-errno -fno-trapping-math -Werror=format -Wno-stringop-overflow -DHAVE_SVE_CPU_DEFINITION -DHAVE_SVE256_CPU_DEFINITION -O3 -DNDEBUG -DNDEBUG -std=gnu++17 -fPIC -DTORCH_USE_LIBUV -DCAFFE2_USE_GLOO -D__NEON__ -Wall -Wextra -Wdeprecated -Wno-unused-parameter -Wno-missing-field-initializers -Wno-type-limits -Wno-array-bounds -Wno-unknown-pragmas -Wno-strict-overflow -Wno-strict-aliasing -Wunused-function -Wunused-variable -Wunused-but-set-variable -Wno-maybe-uninitialized -fvisibility=hidden -O2 -pthread -fopenmp -march=armv8.2-a+fp16 -MD -MT caffe2/CMakeFiles/torch_cpu.dir/__/aten/src/ATen/native/BlasKernel.cpp.o -MF caffe2/CMakeFiles/torch_cpu.dir/__/aten/src/ATen/native/BlasKernel.cpp.o.d -o caffe2/CMakeFiles/torch_cpu.dir/__/aten/src/ATen/native/BlasKernel.cpp.o -c /home/abhishekk/org11/venv/pytorch/aten/src/ATen/native/BlasKernel.cpp
> /home/abhishekk/org11/venv/pytorch/aten/src/ATen/native/BlasKernel.cpp: In function ‘float at::native::blas_impl::{anonymous}::dot_with_fp32_arith_bfdot(const c10::BFloat16*, const c10::BFloat16*, int64_t)’:
> /home/abhishekk/org11/venv/pytorch/aten/src/ATen/native/BlasKernel.cpp:582:25: error: inlining failed in call to ‘always_inline’ ‘float at::native::blas_impl::{anonymous}::dot_with_fp32_arith_tail_after_main_loop(const T*, const T*, int64_t, float) [with T = c10::BFloat16]’: target specific option mismatch
>   582 | C10_ALWAYS_INLINE float dot_with_fp32_arith_tail_after_main_loop(
>       |                         ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
> /home/abhishekk/org11/venv/pytorch/aten/src/ATen/native/BlasKernel.cpp:610:50: note: called from here
>   610 |   return dot_with_fp32_arith_tail_after_main_loop(vec1, vec2, len, reducedSum);
>       |          ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~
> /home/abhishekk/org11/venv/pytorch/aten/src/ATen/native/BlasKernel.cpp:582:25: error: inlining failed in call to ‘always_inline’ ‘float at::native::blas_impl::{anonymous}::dot_with_fp32_arith_tail_after_main_loop(const T*, const T*, int64_t, float) [with T = c10::BFloat16]’: target specific option mismatch
>   582 | C10_ALWAYS_INLINE float dot_with_fp32_arith_tail_after_main_loop(
>       |                         ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
> /home/abhishekk/org11/venv/pytorch/aten/src/ATen/native/BlasKernel.cpp:610:50: note: called from here
>   610 |   return dot_with_fp32_arith_tail_after_main_loop(vec1, vec2, len, reducedSum);
>       |          ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~
> ```

this looks like a GCC bug: it's complaining it can't inline a function without a target attribute into a function with a target attribute. doesn't make sense to me why that wouldn't be OK. here is clang accepting it and gcc not accepting it: https://godbolt.org/z/3Wj1K1nPa looks like it's required that the `-march` flag is set and doesn't match the bf16 arch flag for the bug to be triggered. Unfortunately, the issue repros all the way up to GCC trunk.

**Comment by swolchok** [comment]
I will probably reapply this and just disable the BF16 option for GCC.

**Comment by abhishek-iitmadras** [comment]
> I will probably reapply this and just disable the BF16 option for GCC.

Hi @swolchok (cc @malfet )

If I add below code:
```
TARGET_ARM_BF16_ATTRIBUTE
static C10_ALWAYS_INLINE float dot_with_fp32_arith_tail_after_main_loop_bf16(
    const BFloat16* vec1,
    const BFloat16* vec2,
    int64_t len,
    float reducedSum) {
  const auto len_aligned = len & ~(kF32ElementsPerIteration - 1);
  for (int j = len_aligned; j < len; ++j) {
    reducedSum += static_cast<float>(vec1[j]) * static_cast<float>(vec2[j]);
  }
  return reducedSum;
}
```

And modified `dot_with_fp32_arith_bfdot` to use `dot_with_fp32_arith_tail_after_main_loop_bf16` specific to bf16 then build pass successfully.

```
#if COMPILER_SUPPORTS_BF16_TARGET
TARGET_ARM_BF16_ATTRIBUTE float
dot_with_fp32_arith_bfdot(const BFloat16* vec1, const BFloat16* vec2, int64_t len) {
  auto reducedSum = dot_with_fp32_arith_main_loop_bfdot(vec1, vec2, len);
  return dot_with_fp32_arith_tail_after_main_loop_bf16(vec1, vec2, len, reducedSum);
}
#endif // COMPILER_SUPPORTS_BF16_TARGET
```

So the issue here might me like in function `dot_with_fp32_arith_bfdot`, You try to return `dot_with_fp32_arith_tail_after_main_loop` and in `dot_with_fp32_arith_tail_after_main_loop` , you use `dot_with_fp32_arith_vectorized_tail_inner_loop` which is not specific to bf16.

Correct me if I am wrong.

**Comment by swolchok** [comment]
> you use dot_with_fp32_arith_vectorized_tail_inner_loop which is not specific to bf16.
>
> Correct me if I am wrong.

It should be fine to inline a not-specific-to-bf16 function in a specific-to-bf16 function. In fact, clang does it correctly. I would prefer not to duplicate code to work around compiler bugs.

**Comment by swolchok** [comment]
> prefer not to duplicate code

it will be slightly ugly, but we can put the body of the one function we'd need to duplicate in a macro so as not to punish people for using ARM GCC. I will try that.

**Comment by swolchok** [comment]
#137377 is the second attempt
